EDA_functions/config_explanation.md
# Объяснение параметров конфига

## `DATA_FILE_PATH`
Укажите относительный путь до файла с данными (сделан по умолчанию),
либо абсолютный путь до файла.

## `CLEANING_CONFIG`
- ### remove_rows_with_values 
Удаление строк с определёнными значениями. Укажите название столбца и значение,
по которому будет происходить удаление.  
Формат: `"Название_столбца": ["Значение_1", "Значение_2"]`
- ### custom_conditions
Удаление строк по своим, определённым условиям. Укажите название столбца и условие для удаления.  
Формат: ``" `Название_столбца` условие"``
- ### required_columns
Удаление строк с пустыми значениями в данных столбцах. Укажите название столбцов, в которых
обязательно должны присутствовать данные.  
Формат: `["Столбец_1", "Столбец_2"]`

## `PROPHET_CONFIG`

- ### model_params
Параметры обучения модели. Укажите:
1. `seasonality_mode` ('additive'/'multiplicative') - Как модель рассматривает сезонные колебания. 
    - **Additive**: Сезонные колебания имеют примерно постоянную амплитуду во времени. *Пример: каждый декабрь продажи увеличиваются на +1 млн руб.*  
   - **Multiplicative**: Амплитуда сезонных колебаний растет/падает вместе с трендом. *Пример: каждый декабрь продажи увеличиваются на +20%.*  
2. `yearly_seasonality` (True/False) - Включает годовую сезонность (период = 365.25 дней).
    - **True**: Модель будет искать повторяющиеся паттерны в течение года. Пример: Пики продаж в декабре, спад в январе, сезонность в туризме и т.д
    - **False**: Игнорирует годовую сезонность.
3. `weekly_seasonality` (True/False) - Включает недельную сезонность (период = 7 дней).
   - **True**: Модель будет искать паттерны, повторяющиеся каждую неделю. Пики продаж в пятницу-субботу, спад в понедельник-вторник, низкая активность в воскресенье.
   - **False**: Игнорирует недельную сезонность.
4. `daily_seasonality` (True/False) - Включает дневную сезонность (период = 1 день).
   - **True**: Модель будет искать паттерны в течение дня. Пик посещаемости сайта с 12:00 до 15:00, продажи кофе утром и вечером.
   - **False**: Игнорирует дневную сезонность. Для больших данных по продажам - не нужный фактор, всегда отключайте.
5. `changepoint_prior_scale` [0.001, 1.0] - Контролирует гибкость тренда. Это сила, с которой модель может изменять направление тренда в "точках изменения" (changepoints).
   - **0.001 - 0.01**: Очень низкие значения. Тренд будет жестким, линейным, почти негнущимся. Модель будет игнорировать мелкие изменения и шумы, следуя общему направлению. Риск: не дообучение.
   - **0.05 - 0.2**: "Золотая середина" и самый распространенный диапазон для старта. Тренд сможет плавно изгибаться, адаптируясь к основным изменениям в данных.
   - **0.3 - 1.0+**: Очень высокие значения. Тренд станет чрезвычайно гибким и "нервным" - будет пытаться подстроиться под каждое мелкое колебание в исторических данных. Риск: переобучение.
6. `seasonality_prior_scale` [0.01, 100] - Контролирует силу сезонных компонентов (и годовой, и недельной).
   - **0.01 - 1.0**: Низкие значения. Сила сезонности/эффекта праздников сильно регулируется (сглаживается). Модель будет считать, что сезонные эффекты выражены слабо.
   - **5.0 - 20.0**: Значения по умолчанию и стандартный диапазон. (seasonality_prior_scale по умолчанию = 10.0). Модель доверяет данным и строит сезонность с умеренной силой.
   - **50.0 - 100.0+**: Очень высокие значения. Модель будет полностью доверять историческим данным в плане сезонности. Это может привести к резким, почти "пилообразным" сезонным компонентам, которые уловят даже случайные выбросы.
7. `holidays_prior_scale` [0.01, 100] - Контролирует силу эффекта праздников (например, Новый год, 8 марта и т.д.), которые вы добавили через add_country_holidays.  
Те же самые настройки и значения, что и в `seasonality_prior_scale`, 
однако `holidays_prior_scale` обычно требуется меньше, чем `seasonality_prior_scale`, потому что праздников меньше и их эффект, как правило, более локализован во времени.

- ### forecast_params
Параметры прогнозирования. Укажите:
1. `periods` (число) - количество периодов, на которые будет делаться прогноз. 
   - **7 - 30**: краткосрочный прогноз (на неделю/месяц). Наиболее точный.
   - **31 - 90**: среднесрочный прогноз (до квартала). Оптимально для бизнес-планирования.
   - **91 - 365**: долгосрочный прогноз (до года). НаиМ=менее точный, но полезен для стратегического планирования.
2. `freq` (D/W/M/Q/Y) - определяет единицу измерения для периодов прогнозирования.  
Соответственно дни, недели, месяцы, кварталы, годы. Частота должна совпадать с частотой ваших исходных данных! Если у вас продажи по дням, прогнозируйте по дням ('D').
3. `country_holidays` (Код страны) - определяет праздники по коду страны;
4. `custom_holidays` - праздники и события самой компании ("черная пятница", технические сбои, продажи нового продукта), которые повлияли на продажи.
Формат: `pd.Dataframe({
'holiday': название_события
'ds': pd.to_datetime(['ГГГГ-ММ-ДД', 'Дата_2', 'Дата_3', ...]),  
'lower_window': число_дней_ДО_события,              
'upper_window': число_дней_ПОСЛЕ_события,
})`.  
Например, "lower_window: 2" и "upper_window: 6" будет означать, что событие влияло на статистику в течение 9 дней
(2 дня до начала + день начала + 6 дней после начала). В поле даты записывайте все даты, в которое произошло событие
(то есть если чёрная пятница была в последнюю пятницу ноября каждого года, а вы взяли период за 2 года - запишите две даты данных пятниц
за первый и второй год).


EDA_functions/config.py
# Конфигурационный файл для настройки обработки данных

# Путь к файлу с данными
# DATA_FILE_PATH = "C:/Users/Joker/PycharmProjects/Practice/Data/sales_data.xlsx"
DATA_FILE_PATH = "../Data/sales_data.xlsx"

# Настройки очистки данных
CLEANING_CONFIG = {
    # Удаление строк с определенными значениями в столбцах
    "remove_rows_with_values": {
        # "Регион": ["Кострома", "test", "TEST"],
        # "Категория": ["Тест", "test"],
        # "Продукт": ["тестовый", "test"]
    },

    # Дополнительные условия для удаления строк
    "custom_conditions": [
        # Пример: удалить строки, где количество больше 1000
        # "`Кол-во` > 1000",
        # Пример: удалить строки, где сумма меньше 10
        # "`Сумма` < 10"
    ],

    # Столбцы, которые не должны содержать пустые значения
    "required_columns": ["Дата продажи", "Кол-во", "Сумма"]
}

# Конфигурация для модели Prophet
PROPHET_CONFIG = {
    "model_params": {
        "seasonality_mode": "multiplicative",  # 'additive' or 'multiplicative'
        "yearly_seasonality": True,
        "weekly_seasonality": True,
        "daily_seasonality": False,
        "changepoint_prior_scale": 0.05,
        "seasonality_prior_scale": 10.0,
        "holidays_prior_scale": 10.0,
    },
    "forecast_params": {
        "periods": 90,  # Прогноз на 90 дней вперед
        "freq": 'D',     # Частота прогноза ('D' - день, 'W' - неделя, 'M' - месяц)
    },
    "country_holidays": "RU",  # Код страны для автоматического добавления праздников (Russia)
    # Полный список: https://github.com/dr-prodigy/python-holidays
    "custom_holidays": [
        # pd.DataFrame({
        # 'holiday': 'promo_week', # название события
        # 'ds': pd.to_datetime(['2023-02-01', '2023-08-01']), # Даты начала события
        # 'lower_window': 0, # За сколько дней ДО события начинается эффект
        # 'upper_window': 6, # Сколько дней ПОСЛЕ начала события длится эффект (то есть продолжительность)
        # }),
    ]
}